### PR TITLE
Accept ostree commit and container container image input

### PR DESCRIPTION
**Pull request by lukewarmtemp**
Allow both ostree commits and container images as input for the `org.osbuild.ostree.deploy` stage.

### CODE REVIEW

File: test/data/manifests/fedora-coreos-container.mpp.yaml [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+          #   origin: org.osbuild.source
+          #   references:
+          #     "2c39e9b6a6b26226b626f4daf815d5307a96b2cd57463e94f7c618e312a61980":
+          #       ref: fedora/x86_64/coreos/stable
+      - type: org.osbuild.ostree.selinux
+        options:
+          deployment:
+            osname: fedora-coreos
+            ref: localhost/fcos
+          container: true
```
Comment (lukewarmtemp) [inline test/data/manifests/fedora-coreos-container.mpp.yaml:102]: Maybe look into a better way to distinguish between ostree commit and container image inputs than having a `container` boolean?

File: test/data/manifests/fedora-coreos-container.mpp.yaml [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+      - type: org.osbuild.mkdir
+        options:
+          paths:
+            - path: /boot/efi
+              mode: 448
+      - type: org.osbuild.ignition
+      - type: org.osbuild.ostree.deploy
+        options:
+          osname: fedora-coreos
+          ref: localhost/fcos
```
Comment (lukewarmtemp) [inline test/data/manifests/fedora-coreos-container.mpp.yaml:66]: Might be redundant considering how the container image input defined below also specifies the ref. I'm not sure if we should make this optional though since we still have the case with no explicit inputs to the deploy stage (ie. the old example using the `org.osbuild.ostree.pull` stage). However, I think it just makes it slightly more confusing when you need to specify the same ref in two places.

File: test/data/manifests/fedora-coreos-container.mpp.yaml (hunk 1 above)
Comment (dustymabe) [reply test/data/manifests/fedora-coreos-container.mpp.yaml:102]: We don't need `container: true`. The first deployment (and we'll only ever have one deployment anyway) in ostree will always be `ref: ostree/1/1/0` so just change ref to `ref: ostree/1/1/0` here. Should work and then you can delete all the hacks you added for `--write-commit-to` and stuff.

File: tools/osbuild-mpp [hunk 3]
```diff
@@ -1498,10 +1498,15 @@ class ManifestFileV2(ManifestFile):
             embed_data(ip, mpp)
 
     def _process_container(self, stage):
-        if stage.get("type", "") != "org.osbuild.skopeo":
+        if stage.get("type", "") not in ["org.osbuild.skopeo", "org.osbuild.ostree.deploy"]:
             return
 
         inputs = element_enter(stage, "inputs", {})
+
+        # prevents an empty "images" object from being created when an ostree commit is used as an input and not a container image
```
Comment (dustymabe) [inline tools/osbuild-mpp:1506]: A little more context might be nice:

```suggestion
        # The ostree.deploy stage accepts both containers or
        # ostree commits as input. If this is an ostree.deploy
        # stage and there is no images in the inputs then let's
        # return early. This prevents an empty "images" object
        # from being created when an ostree commit is used as
        # an input and not a container image     
```

File: test/data/manifests/fedora-coreos-container.mpp.yaml (hunk 1 above)
Comment (lukewarmtemp) [reply test/data/manifests/fedora-coreos-container.mpp.yaml:102]: Using `ref: ostree/1/1/0` should work for both ostree commit and container image inputs. I just get a few warning outputs from using my actual system as the build root that mess with the output of `ostree rev-parse`, so I can't actually test the container input case (I'm using a custom `ostree` and `rpm-ostree` binary). However, by printing the output I can confirm that the parsed refspec is correct.

I can retest this once the new release of `ostree-rs-ext` is out.

Comment (dustymabe) [reply test/data/manifests/fedora-coreos-container.mpp.yaml:102]: > so I can't actually test the container input case

I just tested it - seemed to work.

File: stages/org.osbuild.ostree.selinux [hunk 4]
```diff
@@ -37,11 +37,14 @@ SCHEMA = """
         "default": 0
       }
     }
+  },
+  "container": {
+    "description": "Bool to indicate if the deployment is a container",
+    "type": "boolean"
```
Comment (dustymabe) [inline stages/org.osbuild.ostree.selinux:43]: delete

File: stages/org.osbuild.ostree.selinux [hunk 5]
```diff
@@ -51,7 +54,7 @@ def main(tree, options):
     # this created a state root at `osname`
     stateroot = f"{tree}/ostree/deploy/{osname}"
 
-    root = ostree.deployment_path(tree, osname, ref, serial)
+    root = ostree.deployment_path(tree, osname, ref, serial)
```
Comment (dustymabe) [inline stages/org.osbuild.ostree.selinux:57]:
```suggestion
    root = ostree.deployment_path(tree, osname, ref, serial)
```

File: test/data/manifests/fedora-coreos-container.mpp.yaml (hunk 2 above)
Comment (dustymabe) [reply test/data/manifests/fedora-coreos-container.mpp.yaml:66]: I agree. I think we can make this optional in the schema, but add a check for it in the code. i.e. validation won't catch it but at least the code would.

File: stages/org.osbuild.ostree.deploy [hunk 6]
```diff
@@ -112,39 +126,86 @@ def main(tree, inputs, options):
...
+      
+      elif input_type == 'images':
+        images = containers.parse_containers_input(inputs)
+        for i in inputs["images"]:
+            print(i)
+
+        for image in images.values():
+          with containers.container_source(image) as (image_name, image_source):
+            extra_args = []
+            imgref = f"{remote}:{image_source}"
```
Comment (dustymabe) [inline stages/org.osbuild.ostree.deploy:186]: we probably shouldn't rely on `remote` here

```suggestion
            imgref = f"ostree-unverified-image:{image_source}"
```

File: stages/org.osbuild.ostree.deploy [hunk 7]
```diff
@@ -112,39 +126,86 @@ def main(tree, inputs, options):
...
+
+            if remote:
```
Comment (dustymabe) [inline stages/org.osbuild.ostree.deploy:188]: should the following block be conditional on `if remote:`?

File: stages/org.osbuild.ostree.deploy (hunk 6 above)
Comment (lukewarmtemp) [reply stages/org.osbuild.ostree.deploy]: Do we want the ostree commit input case to also not rely on remote, just like how we changed the container image input case to not rely on remote (ie. `ref = f"fedora:{ref}"`)?

File: test/data/manifests/fedora-coreos-container.mpp.yaml (hunk 2 above)
Comment (lukewarmtemp) [reply test/data/manifests/fedora-coreos-container.mpp.yaml]: Should we make `ref` optional for the ostree commit input case too? I think we should be able to get the ref from the inputs object:
```
print(inputs)
{'commits': {'path': '/run/osbuild/inputs/commits/repo', 'data': {'refs': {'2c39e9b6a6b26226b626f4daf815d5307a96b2cd57463e94f7c618e312a61980': {'ref': 'fedora/x86_64/coreos/stable'}}}}}
```

File: stages/org.osbuild.ostree.deploy (hunk 6 above)
Comment (lukewarmtemp) [reply stages/org.osbuild.ostree.deploy]: Or maybe we should keep using remote since we still need to be able to accept an ostree commit created from a previous stage in the pipeline which might have a remote different from the `fedora` remote?

File: stages/org.osbuild.ostree.deploy [hunk 8]
```diff
@@ -84,12 +85,33 @@ SCHEMA_2 = """
...
+    "required": ["images"]
+  }, {
+      "not": { 
+        "anyOf": [{
+            "required": ["commits"]
+          }, {
+            "required": ["images"]
+          }] 
+      }
+  }],
```
Comment (dustymabe) [inline stages/org.osbuild.ostree.deploy:101]: I'm trying to understand the `not/anyOf` part. Can you dumb it down for me?

Comment (lukewarmtemp) [reply stages/org.osbuild.ostree.deploy:101]: When I didn't put in any inputs (so that it matches the old example using the `org.osbuild.ostree.pull` stage), I got a schema error since it didn't match either having a `commits` or `images` input. This is the logic for the three statements you've highlighted:

```
# Case 1: We have a commits input
"required": ["commits"] = True
"required": ["images"] = False
"not/anyOf" = ! ( True || False ) = False

# Case 2: We have an images input
"required": ["commits"] = False
"required": ["images"] = True
"not/anyOf" = ! ( True || False ) = False

# Case 3: We have neither a commits or images input
"required": ["commits"] = False
"required": ["images"] = False
"not/anyOf" = ! ( False || False ) = True
```
As seen above, each of the cases only results in one True statement which is needed to pass the `oneOf` statement (basically an XOR).

Comment (lukewarmtemp) [reply stages/org.osbuild.ostree.deploy:101]: In addition, we want it to fail if we specify more than one input as well:
```
# Case 4: We have both a commits or images input
"required": ["commits"] = True
"required": ["images"] = True
"not/anyOf" = ! ( True || True ) = False
```
Which fails the XOR since it results in two `True` statements

Comment (dustymabe) [reply stages/org.osbuild.ostree.deploy:101]: 👍

File: stages/org.osbuild.ostree.deploy [hunk 9]
```diff
@@ -100,51 +122,96 @@ def make_fs_identifier(desc):
...
+    input_type = None
+    for key in inputs:
+      input_type = key
+    
+    # check if "ref" key exists to avoid accessing a non-existant key, value pair
+    if "ref" not in options.keys():
+      if not input_type or input_type == "commits":
+        raise ValueError("ref missing")
+    else:
+      ref = options["ref"]
```
Comment (dustymabe) [inline stages/org.osbuild.ostree.deploy:140]: I think this is just:

```
ref = options.get("ref")
```

i.e. if it's not in the options it won't error.

Comment (dustymabe) [reply stages/org.osbuild.ostree.deploy:140]: oh, I see. you want to add some verification that if it is needed we'll raise a valueError if it wasn't provided.

Comment (lukewarmtemp) [reply stages/org.osbuild.ostree.deploy:140]: Yeah, the main point of this part is to add verification in the code since we made `ref` optional in the schema, but using `ref = options.get("ref")`and checking for `None` is still better than using `"ref" not in options.keys()`

File: test/data/manifests/fedora-coreos-container.mpp.yaml (hunk 2 above)
Comment (lukewarmtemp) [reply test/data/manifests/fedora-coreos-container.mpp.yaml]: After discussion, decided to make ref a requirement for ostree commit inputs still.
